velox/common/memory/tests/SharedArbitratorTest.cpp

```diff
@@ -958,5 +960,1636 @@ TEST_F(MockSharedArbitrationTest, concurrentArbitrationWithTransientRoots) {
+    std::shared_ptr<core::QueryCtx> aggregationQueryCtx;
+    if (sameQuery) {
+      aggregationQueryCtx = fakeMemoryQueryCtx;
+    } else {
+      aggregationQueryCtx = newQueryCtx(kMemoryCapacity);
+    }
+
+    folly::EventCount fakeAllocationWait;
+    auto fakeAllocationWaitKey = fakeAllocationWait.prepareWait();
+
+    const auto fakeAllocationSize = kMemoryCapacity;
+
+    std::atomic<bool> injectAllocationOnce{true};
+    fakeOperatorFactory_->setAllocationCallback([&](Operator* op) {
+      if (!injectAllocationOnce.exchange(false)) {
+        return Allocation{};
+      }
+      fakeAllocationWait.wait(fakeAllocationWaitKey);
+      auto buffer = op->pool()->allocate(fakeAllocationSize);
+      return Allocation{buffer, fakeAllocationSize};
+    });
+
+    std::thread aggregationThread([&]() {
+      auto task =
+          AssertQueryBuilder(duckDbQueryRunner_)
+              .queryCtx(aggregationQueryCtx)
+              .plan(PlanBuilder()
+                        .values(vectors)
+                        .singleAggregation({"c0", "c1"}, {"array_agg(c2)"})
+                        .planNode())
+              .assertResults(
+                  "SELECT c0, c1, array_agg(c2) FROM tmp GROUP BY c0, c1");
+      waitForTaskCompletion(task.get());
+      fakeAllocationWait.notify();
+    });
+
+    std::thread memThread([&]() {
+      auto task =
+          AssertQueryBuilder(duckDbQueryRunner_)
+              .queryCtx(fakeMemoryQueryCtx)
+              .plan(PlanBuilder()
+                        .values(vectors)
+                        .addNode([&](std::string id, core::PlanNodePtr input) {
+                          return std::make_shared<FakeMemoryNode>(id, input);
+                        })
+                        .planNode())
+              .assertResults("SELECT * FROM tmp");
+    });
+
+    aggregationThread.join();
+    memThread.join();
+    Task::testingWaitForAllTasksToBeDeleted();
+  }
+}
+
+DEBUG_ONLY_TEST_F(SharedArbitrationTest, reclaimFromJoinBuilder) {
+  const int numVectors = 32;
+  std::vector<RowVectorPtr> vectors;
+  for (int i = 0; i < numVectors; ++i) {
+    vectors.push_back(newVector());
+  }
+  createDuckDbTable(vectors);
+  std::vector<bool> sameQueries = {false, true};
+  for (bool sameQuery : sameQueries) {
+    SCOPED_TRACE(fmt::format("sameQuery {}", sameQuery));
+    const auto spillDirectory = exec::test::TempDirectoryPath::create();
+    std::shared_ptr<core::QueryCtx> fakeMemoryQueryCtx =
+        newQueryCtx(kMemoryCapacity);
+    std::shared_ptr<core::QueryCtx> joinQueryCtx;
+    if (sameQuery) {
+      joinQueryCtx = fakeMemoryQueryCtx;
+    } else {
+      joinQueryCtx = newQueryCtx(kMemoryCapacity);
+    }
+
+    folly::EventCount fakeAllocationWait;
+    auto fakeAllocationWaitKey = fakeAllocationWait.prepareWait();
+    folly::EventCount taskPauseWait;
+    auto taskPauseWaitKey = taskPauseWait.prepareWait();
+
+    const auto joinMemoryUsage = 32L << 20;
```

Comment:
Could you explain why joinMemoryUsage is equal to 32M here? And why does fakeAllocationSize need to be set to 512M - 32M / 2?